### PR TITLE
Feat : `TextField` cursor color matching M2 and M3 Spec in error state

### LINKED ISSUE #113750: TextField cursor color does not match state spec in error state
`TextField`s in error states should have a cursor color which matches other error themed colors:

Material 2 Spec:
34VHXF3dJcX4VzJ (1)

Material 3 Spec:
AUuETGJ23o4Gs6m

However, this is not set in the Material theme and falls back to [`colorScheme.primary`]:(https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/app.dart#L931) 

Current implementation:
5yZF8fphsHEkqET

The `TextField` widget derives its `cursorColor` property from itself, or defers to a `TextSelectionTheme`. The `InputDecoration` which supplies themed properties to the `TextField` does not have a `selectionTheme` field. I would propose adding it to `InputDecoration`.

### PR DESCRIPTION
**Pull request by hasnentai**
`TextField` in error states should have a cursor color which matches other error-themed colours.

Fixes: #113750
Related: https://github.com/flutter/flutter/issues/114950

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
Could you add the newline back at the end of text_field.dart? The analyzer is complaining (https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8790918503105134049/+/u/Analyze/analyze/test_stdout):    
> info • Missing a newline at the end of the file • packages/flutter/lib/src/material/text_field.dart:1465:46 • eol_at_end_of_file

**Comment by hasnentai** [comment]
I was wondering why the checks were failing. Thanks 
I have added a new line in the file

File: packages/flutter/test/material/text_form_field_test.dart [hunk 1]
```diff
@@ -1169,4 +1169,32 @@ void main() {
...
   });
+
+
+    testWidgets('Error color for cursor while validation', (WidgetTester tester) async {
+
+    const  Color errorColor = Color(0xff123456);
+
+    await tester.pumpWidget(
+      MaterialApp(
+        theme: ThemeData(colorScheme:  const ColorScheme.light(error: errorColor)),
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1180]: I think there are multiple spaces before the `const`.

Also nit: Maybe split the colorScheme parameter onto its own line.

File: packages/flutter/lib/src/material/text_field.dart [hunk 2]
```diff
@@ -1230,7 +1243,7 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
         textSelectionControls ??= cupertinoTextSelectionHandleControls;
         paintCursorAboveText = true;
         cursorOpacityAnimates = true;
-        cursorColor = widget.cursorColor ?? selectionStyle.cursorColor ?? cupertinoTheme.primaryColor;
+        cursorColor = _getCursorColor(selectionStyle, theme);
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1246]: Should `theme` be `cupertinoTheme` here?

File: packages/flutter/lib/src/material/text_field.dart [hunk 3]
```diff
@@ -1243,7 +1256,7 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
         textSelectionControls ??= cupertinoDesktopTextSelectionHandleControls;
         paintCursorAboveText = true;
         cursorOpacityAnimates = false;
-        cursorColor = widget.cursorColor ?? selectionStyle.cursorColor ?? cupertinoTheme.primaryColor;
+        cursorColor = _getCursorColor(selectionStyle, theme);
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1259]: Same thing about cupertinoTheme here too.

File: packages/flutter/lib/src/material/text_field.dart [hunk 4]
```diff
@@ -1149,6 +1151,17 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
     }
   }
 
+  /// Sets the cursor color to error color from decoration or from theme colorScheme if [_hasError] is true else
+  /// it will return the color from [widget.cursorColor] param if not null. if [widget.cursorColor] is null then it will return default style else return
+  /// color from theme colorScheme.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1156]: Really I think this entire comment is self explanatory by looking at the code and the method signature.  I would just remove it.

File: packages/flutter/lib/src/material/text_field.dart [hunk 5]
```diff
@@ -1149,6 +1151,17 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
...
+  Color _getCursorColor(DefaultSelectionStyle selectionStyle, ThemeData theme){
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1157]:
```suggestion
  Color _getCursorColor (DefaultSelectionStyle selectionStyle, ThemeData theme) {
```

File: packages/flutter/lib/src/material/text_field.dart [hunk 6]
```diff
@@ -1149,6 +1151,17 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
...
+    if(_hasError){
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1158]:
```suggestion
    if (_hasError) {
```

File: packages/flutter/lib/src/material/text_field.dart [hunk 7]
```diff
@@ -1149,6 +1151,17 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
...
+      return _errorColor;
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:1159]: Nit: You could just inline this here rather than making it a getter, but either way works.

**Review by justinmc — COMMENTED** [review]
Thanks for fixing this!  Mostly just minor comments below.

File: packages/flutter/test/material/text_form_field_test.dart [hunk 8]
```diff
@@ -1169,4 +1169,32 @@ void main() {
     final EditableText editableText = tester.widget(find.byType(EditableText));
     expect(editableText.magnifierConfiguration, equals(myTextMagnifierConfiguration));
   });
+
+
+    testWidgets('Error color for cursor while validation', (WidgetTester tester) async {
```
[the remark this answers is not in the document]
Comment (hasnentai) [reply packages/flutter/test/material/text_form_field_test.dart:1174]: ahh sorry, I just missed it. I have made the changes

File: packages/flutter/test/material/text_form_field_test.dart [hunk 9]
```diff
@@ -1169,4 +1169,32 @@ void main() {
...
+            ),
+          ),
+        ),
+      ),
+    );
+    await tester.enterText(find.byType(TextField), 'a');
+    final EditableText textField =  tester.widget(find.byType(EditableText).first);
+    await tester.pump();
+    expect(textField.cursorColor, errorColor);
+
```
[the remark this answers is not in the document]
Comment (hasnentai) [reply packages/flutter/test/material/text_form_field_test.dart:1198]: Done

File: packages/flutter/test/material/text_form_field_test.dart (hunk 1 above)
Comment (hasnentai) [reply packages/flutter/test/material/text_form_field_test.dart:1180]: I have made it to a newline of color scheme to its own

File: packages/flutter/lib/src/material/text_field.dart (hunk 2 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1246]: Great catch, Added Cupertino Theme for Mac and Ios platform case

File: packages/flutter/lib/src/material/text_field.dart (hunk 3 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1259]: done

File: packages/flutter/lib/src/material/text_field.dart (hunk 4 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1156]: Removed

File: packages/flutter/lib/src/material/text_field.dart (hunk 5 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1157]: Added space

File: packages/flutter/lib/src/material/text_field.dart (hunk 6 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1158]: Added space

File: packages/flutter/lib/src/material/text_field.dart (hunk 7 above)
Comment (hasnentai) [reply packages/flutter/lib/src/material/text_field.dart:1159]: I have made it inline

**Comment by hasnentai** [comment]
Hi @justinmc Google Test is failing 
What could be the possible reason ?

**Comment by justinmc** [comment]
Hmm it's not showing up in the Google testing tool for me.  Can you try pushing a merge commit with master?

**Comment by hasnentai** [comment]
@justinmc Finally the checks are passing 😀. Please review it once

File: packages/flutter/test/material/text_form_field_test.dart [hunk 10]
```diff
@@ -1169,4 +1169,28 @@ void main() {
     final EditableText editableText = tester.widget(find.byType(EditableText));
     expect(editableText.magnifierConfiguration, equals(myTextMagnifierConfiguration));
   });
+
+  testWidgets('Error color for cursor while validation', (WidgetTester tester) async {
+    const Color errorColor = Color(0xff123456);
+    await tester.pumpWidget(MaterialApp(
+      theme: ThemeData(
+        colorScheme: const ColorScheme.light(error: errorColor)
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1177]: Missing a trailing comma here.

File: packages/flutter/test/material/text_form_field_test.dart [hunk 11]
```diff
@@ -1169,4 +1169,28 @@ void main() {
     final EditableText editableText = tester.widget(find.byType(EditableText));
     expect(editableText.magnifierConfiguration, equals(myTextMagnifierConfiguration));
   });
+
+  testWidgets('Error color for cursor while validation', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1173]: Nit: "while validation" => "during validation" or "while validating"

File: packages/flutter/test/material/text_form_field_test.dart [hunk 12]
```diff
@@ -1169,4 +1169,28 @@ void main() {
...
+    const Color errorColor = Color(0xff123456);
+    await tester.pumpWidget(MaterialApp(
+      theme: ThemeData(
+        colorScheme: const ColorScheme.light(error: errorColor)
+      ),
+      home: Material(
+        child: Center(
+          child: TextFormField(
+            enabled: true,
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1182]: Is this needed?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I've triggered Google tests for this PR, so let's see what that says too before merge.

File: packages/flutter/test/material/text_form_field_test.dart (hunk 11 above)
Comment (hasnentai) [reply packages/flutter/test/material/text_form_field_test.dart]: Updated the desc

File: packages/flutter/test/material/text_form_field_test.dart (hunk 10 above)
Comment (hasnentai) [reply packages/flutter/test/material/text_form_field_test.dart]: Added two commos one after `errorColor` and one after `)` so now its like this 
```dart
        colorScheme: const ColorScheme.light(
          error: errorColor,
        ),
```

**Comment by hasnentai** [comment]
Hi @justinmc, I am not sure why Framework test widgets are failing after my last commit. Even I reverted to commit which was passing all the checks but it's failing now. I even checked the logs but was not able to understand. It would be great if you can provide some input on the same

**Comment by justinmc** [comment]
The Google test failures were all to be expected - they should be accepted with the updated cursor color.  I've set them to show a pass here.

The framework test that's failing is this:

```
The following TestFailure was thrown running a test:
Expected: <null>
  Actual: <33.333333333333336>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///Volumes/Work/s/w/ir/x/w/flutter/packages/flutter/test/widgets/media_query_test.dart:159:5)
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:170:29)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
```

Seems unrelated, right?  I've tried rerunning it.  If that doesn't work, could you push a new merge commit?  If you haven't tried that already.

**Comment by hasnentai** [comment]
I have tried couple of time pushing a new commit on Sunday. May be you can see the history time line of this thread 
But seems to be failing all the time.

**Comment by justinmc** [comment]
Alright let's see if my rerun happens to fix anything, otherwise I'll ask about this test on Discord.

**Comment by justinmc** [comment]
Looks like it passed, maybe a flake.

**Review by justinmc — APPROVED** [review]
Renewing my LGTM.

Also, I forgot but we need a secondary reviewer.  I'll ask around.

**Review by gnprice — APPROVED** [review]
LGTM

**Comment by justinmc** [comment]
Thanks for the secondary review!  I'll plan to merge this tomorrow morning.

**Comment by justinmc** [comment]
@hasnentai Should the handle also be red instead of blue?  I made an issue to track this (https://github.com/flutter/flutter/issues/122538) but I couldn't find a spec or anything that showed it being red.

Screenshot from 2023-03-13 10-29-39

**Comment by justinmc** [comment]
@hasnentai Can you confirm that the handle should be red by default on native Android?  If so then I think we should definitely make the change.

**Comment by justinmc** [comment]
Thanks!

**Pull request merged**